src/processor/fingerprint.cpp

```diff
@@ -298,4 +521,104 @@ std::pair<ddwaf_object, object_store::attribute> http_endpoint_fingerprint::eval
+std::pair<ddwaf_object, object_store::attribute> http_header_fingerprint::eval_impl(
+    const unary_argument<const ddwaf_object *> &headers, ddwaf::timer &deadline) const
+{
+    std::string known_header_bitset;
+    known_header_bitset.resize(standard_headers_length, '0');
+
+    std::string_view user_agent;
+    std::vector<std::string> unknown_headers;
+    std::string normalized_header;
+    for (std::size_t i = 0; i < headers.value->nbEntries; ++i) {
+        if (deadline.expired()) {
+            throw ddwaf::timeout_exception();
+        }
+
+        const auto &child = headers.value->array[i];
+        std::string_view header{
+            child.parameterName, static_cast<std::size_t>(child.parameterNameLength)};
+
+        normalize_header(header, normalized_header);
+        auto [type, index] = get_header_type_and_index(normalized_header);
+        if (type == header_type::standard) {
+            known_header_bitset[index] = '1';
+        } else if (type == header_type::unknown) {
+            unknown_headers.emplace_back(normalized_header);
+        } else if (type == header_type::user_agent && child.type == DDWAF_OBJ_STRING) {
+            user_agent = {child.stringValue, static_cast<std::size_t>(child.nbEntries)};
+        }
+    }
+    std::sort(unknown_headers.begin(), unknown_headers.end());
+
+    auto res =
+        generate_fragment("hdr", string_field{known_header_bitset}, string_hash_field{user_agent},
+            unsigned_field{unknown_headers.size()}, vector_hash_field{unknown_headers});
+
+    return {res, object_store::attribute::none};
+}
+
+// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+std::pair<ddwaf_object, object_store::attribute> http_network_fingerprint::eval_impl(
+    const unary_argument<const ddwaf_object *> &headers, ddwaf::timer &deadline) const
+{
+    std::string ip_origin_bitset;
+    ip_origin_bitset.resize(ip_origin_headers_length, '0');
+
+    unsigned chosen_header = ip_origin_headers_length;
+    std::string_view chosen_header_value;
+    std::string normalized_header;
+    for (std::size_t i = 0; i < headers.value->nbEntries; ++i) {
+        if (deadline.expired()) {
+            throw ddwaf::timeout_exception();
+        }
+
+        const auto &child = headers.value->array[i];
+
+        std::string_view header{
+            child.parameterName, static_cast<std::size_t>(child.parameterNameLength)};
+
+        normalize_header(header, normalized_header);
+        auto [type, index] = get_header_type_and_index(normalized_header);
+        if (type == header_type::ip_origin) {
+            ip_origin_bitset[index] = '1';
+            // Verify not only precedence but also type, as a header of an unexpected
+            // type will be unlikely to be used unless the framework has somehow
+            // broken down the header into constituent IPs
+            if (chosen_header > index && child.type == DDWAF_OBJ_STRING) {
```

Comment:
the chosen header is not necessarily the one with highest precedence. It depends on configured, if the ip is valid, if it is non-public and there other public ones

Reply:
I'm aware but we don't have this information so I went for the standard behaviour. Maybe we can add a new address to propagate this information for the next iteration. Since this fingerprint fragment is.... not very useful yet, I think it should be okay to improve it later.